editor-extensions/vscode/server/src/server.ts

```diff
@@ -0,0 +1,325 @@
+		connection.client.register(DidChangeConfigurationNotification.type, undefined);
+	}
+	if (hasWorkspaceFolderCapability) {
+		connection.workspace.onDidChangeWorkspaceFolders(_event => {
+			connection.console.log('Workspace folder change event received.');
+		});
+	}
+
+	// start the debounce timer
+
+	if (hasConfigurationCapability) {
+		// get the debounce rate from settings
+		connection.workspace.getConfiguration({
+			section: 'grain_language_server'
+		}).then(settings =>
+			debounceTimer = global.setInterval(processChangedDocuments, settings.debounce));
+	} else {
+		debounceTimer = global.setInterval(processChangedDocuments, globalSettings.debounce);
+	};
+});
+
+// The grain cli settings
+interface GrainSettings {
+	maxNumberOfProblems: number;
+	cliPath: string;
+	enableLSP: boolean;
+	trace: string;
+	debounce: number;
+}
+
+// The global settings, used when the `workspace/configuration` request is not supported by the client.
+// Please note that this is not the case when using this server with the client provided in this example
+// but could happen with other clients.
+const defaultSettings: GrainSettings = {
+	maxNumberOfProblems: 1000, cliPath: "grain",
+	enableLSP: true, trace: "off", debounce: 1000
+};
+let globalSettings: GrainSettings = defaultSettings;
+
+// Cache the settings of all open documents
+let documentSettings: Map<string, Thenable<GrainSettings>> = new Map();
+
+connection.onDidChangeConfiguration(change => {
+	connection.console.log("Configuration changed");
+	if (hasConfigurationCapability) {
+		// Reset all cached document settings
+		documentSettings.clear();
+	} else {
+		globalSettings = <GrainSettings>(
+			(change.settings.grain_language_server || defaultSettings)
+		);
+	}
+
+	// clear down the list as we're going to process them all below.
+	changedDocuments = new Set();
+
+	// Revalidate all open text documents
+	documents.all().forEach(doc => validateWithCompiler(doc.uri));
+
+});
+
+function getDocumentSettings(resource: string): Thenable<GrainSettings> {
+	if (!hasConfigurationCapability) {
+		return Promise.resolve(globalSettings);
+	}
+	let result = documentSettings.get(resource);
+	if (!result) {
+		result = connection.workspace.getConfiguration({
+			scopeUri: resource,
+			section: 'grain_language_server'
+		})
+		documentSettings.set(resource, result);
+	}
+	return result;
+}
+
+// Only keep settings for open documents
+// don't process closed documents
+documents.onDidClose(e => {
+	documentSettings.delete(e.document.uri);
+	if (changedDocuments.has(e.document.uri)) changedDocuments.delete(e.document.uri);
+});
+
+// The content of a text document has changed. This event is emitted
+// when the text document first opened or when its content has changed.
+documents.onDidChangeContent(change => {
+
+	// Debounce the actual work as we don't want to run a compile on every keystroke
+
+	changedDocuments.add(change.document.uri);
+
+	//	validateWithCompiler(change.document.uri);
```

Comment:
And it looks like we can drop this now too?
```suggestion
```

Reply:
Cleaned up and deleted